deps/rabbit/src/rabbit_exchange.erl

```diff
@@ -141,6 +142,25 @@ declare(XName, Type, Durable, AutoDelete, Internal, Args, Username) ->
             {ok, X}
     end.
 
+check_exchange_limits(XName) ->
+    Limit = rabbit_misc:get_env(rabbit, exchange_max, infinity),
+    case rabbit_db_exchange:count() >= Limit of
+        false ->
+            ok;
+        true ->
+            case rabbit_db_exchange:exists(XName) of
+                true ->
+                    %% Allow re-declares of existing exchanges when at the
+                    %% exchange limit.
+                    ok;
```

Comment:
Note: allowing redeclares is important because if you reach the exchange limit and restart a Rabbit node, it will fail to boot due to the exchange declarations done in boot steps